numpy/_core/code_generators/cversions.txt

```diff
@@ -76,5 +76,6 @@
 # Version 18 (NumPy 2.0.0)
 0x00000012 = 2b8f1f4da822491ff030b2b37dff07e3
 # Version 19 (NumPy 2.1.0) Only header additions
-# Version 19 (NumPy 2.2.0) No change
 0x00000013 = 2b8f1f4da822491ff030b2b37dff07e3
+# Version 20 (NumPy 2.2.0) Added PyUFunc_FromFuncAndDataAndSignatureAndIdentityAndFlags
```

Comment:
I don't think you should add this function.  Users can already use flags via the new API (yes, it has some other downsides).  And you would have to pass an array of flags anyway.

FWIW, I think you even just guess at `NO_FPEs` if a ufunc loop only contains objects.

Reply:
Is the "new API" you are referring to here the ArrayMethod API? Is there a reference I could read with examples of creating a ufunc using this new API? I gather that this is done with PyUFunc_AddLoopFromSpec(), but this assumes a ufunc already exists, and the method I know to create a ufunc is UFunc_FromFuncAndDataAndSignatureAndIdentity(), which already creates the ArrayMethod internally. What is the method to create a basic ufunc before calling PyUFunc_AddLoopFromSpec()?

Reply:
You can look at the `string_ufuncs` in the `umath` module.  They use this API because they have to.
But I am very happy to hack it here into the old API.  Just not happy to expose it as public API.

Reply:
(And I should say, I am impressed that you dove into it and figured it out!  It isn't common someone dives deep enough to do such thing, even if they are not super complicated by itself)

Reply:
Thanks!

Doing more digging, the new API plan doesn't seem to quite work, at least not if used in the same manner that it is used in string_ufuncs.cpp.

For those ufuncs, it appears that generate_umath.py produces __umath_generated.c which does the initial ufunc creation with the same PyUFunc_FromFuncAndDataAndSignatureAndIdentity() function already being used to create the ufunc for vectorize(). The code in string_ufuncs.cpp loads these already created ufuncs and adds more ArrayMethods to the ufunc->_loops with the new API PyUFunc_AddLoopFromSpec().

If we used that approach in this case, PyUFunc_FromFuncAndDataAndSignatureAndIdentity() would already create the needed ArrayMethod, and trying to add it again except with the correct flags via PyUFunc_AddLoopFromSpec() would fail, as it would be a duplicate (which would be fully blocked by the call to PyUFunc_AddLoop() with ignore_duplicate=0).

Unless there is an existing method to create a ufunc like the ones generated by PyUFunc_FromFuncAndDataAndSignatureAndIdentity() except skipping the step where the ArrayMethod is loaded into _loops, I think the task is to write one, which leads to something quite similar to my first attempted PR.

I think the main problem is just that we don't want to change the public API, right? If so,  just deleting the /*UFUNC_API*/ tag from PyUFunc_FromFuncAndDataAndSignatureAndIdentityAndFlags() ought to do the trick, but for whatever reason, this results in python crashing when a vectorized function is called. I think the problem must have to do with special handling of ufunc_object.c in the build, as moving the function prototype into dispatching.h corrects the problem. I imagine that this is not the right answer, so I'd welcome alternatives and any other feedback.

Reply:
> would already create the needed ArrayMethod

That isn't quite right.  They create a ufunc using that loop, but it would contain no actual loop so there is no `ArrayMethod` being created either.

> don't want to change the public API, right

Yes, that is/was the main issue.  I have to think whether I think still having the helper with flags around is useful if it is used only here.

Reply:
Ok, thanks again. I think I can do better.

Looking more carefully, I now see that the string ufunc initial creation code does not create any actual ArrayMethods because the call to PyUFunc_FromFuncAndDataAndSignatureAndIdentity() passes ntypes=0. We can replicate that approach and then use the contents of PyArray_NewLegacyWrappingArrayMethod as a model for how to build a     PyArrayMethod_Spec to pass to PyUFunc_AddLoopFromSpec(). In addition, ufunc->types and ufunc->ntypes also had to be set manually.

It's been a journey, but a much cleaner implementation now, I think. Please let me know if you see any errors and any other feedback.